ADAL/src/ADInstanceDiscovery.m

```diff
@@ -174,43 +163,97 @@ -(void) validateAuthority: (NSString*) authority
 
 //Checks the cache for previously validated authority.
 //Note that the authority host should be normalized: no ending "/" and lowercase.
--(BOOL) isAuthorityValidated: (NSString*) authorityHost
+- (BOOL)isAuthorityValidated:(NSString *)authorityHost
 {
-    THROW_ON_NIL_EMPTY_ARGUMENT(authorityHost);
-    
-    BOOL validated;
-    @synchronized(self)
+    if (!authorityHost)
     {
-        validated = [mValidatedAuthorities containsObject:authorityHost];
+        return NO;
     }
     
+    BOOL validated = [_validatedAuthorities containsObject:authorityHost];
+    
     NSString* message = [NSString stringWithFormat:@"Checking cache for '%@'. Result: %d", authorityHost, validated];
     AD_LOG_VERBOSE(@"Authority Validation Cache", nil, message);
     return validated;
 }
 
 //Note that the authority host should be normalized: no ending "/" and lowercase.
--(void) setAuthorityValidation: (NSString*) authorityHost
+- (BOOL)addValidAuthority:(NSString *)authorityHost
 {
-    THROW_ON_NIL_EMPTY_ARGUMENT(authorityHost);
-    
-    @synchronized(self)
+    if ([NSString adIsStringNilOrBlank:authorityHost])
     {
-        [mValidatedAuthorities addObject:authorityHost];
+        return NO;
     }
     
+    [_validatedAuthorities addObject:authorityHost];
+    
     NSString* message = [NSString stringWithFormat:@"Setting validation set to YES for authority '%@'", authorityHost];
     AD_LOG_VERBOSE(@"Authority Validation Cache", nil, message);
+    return YES;
+}
+
+- (ADAuthenticationError *)processWebReponse:(ADWebResponse *)webResponse
+                               authorityHost:(NSString *)authorityHost
+                               correlationId:(NSUUID *)correlationId
+{
+    NSInteger code = webResponse.statusCode;
+    if (!(code == 200 || code == 400 || code == 401))
```

Comment:
We should  check if error 400 returns a JSON. If it does not then deserialization would fail